packages/sanity/src/structure/components/requestPermissionDialog/useRoleRequestsStatus.tsx

```diff
@@ -0,0 +1,84 @@
+import {addWeeks, isAfter, isBefore} from 'date-fns'
+import {useEffect, useState} from 'react'
+import {from, of} from 'rxjs'
+import {catchError, map} from 'rxjs/operators'
+import {useClient, useProjectId} from 'sanity'
+
+import {type AccessRequest} from '../../../core/studio/screens'
+
+export const useRoleRequestsStatus = () => {
+  const client = useClient({apiVersion: '2024-07-01'})
+  const projectId = useProjectId()
+  const [status, setStatus] = useState<string>()
+  const [loading, setLoading] = useState(false)
+  const [error, setError] = useState(false)
+
+  useEffect(() => {
+    const checkRoleRequests$ = () => {
+      setLoading(true)
+      if (!client || !projectId) {
+        return of()
+      }
+
+      return from(
+        client.request<AccessRequest[] | null>({
+          url: `/access/requests/me`,
+        }),
+      ).pipe(
+        map((requests) => {
+          if (requests && requests.length) {
+            // Filter requests for the specific project and where type is 'role'
+            const projectRequests = requests.filter(
+              (request) => request.resourceId === projectId && request.type === 'role',
+            )
+
+            const declinedRequest = projectRequests.find((request) => request.status === 'declined')
+            if (declinedRequest) {
+              return 'denied'
+            }
+
+            const pendingRequest = projectRequests.find(
+              (request) =>
+                request.status === 'pending' &&
+                isAfter(addWeeks(new Date(request.createdAt), 2), new Date()),
+            )
+            if (pendingRequest) {
+              return 'pending'
+            }
+
+            const oldPendingRequest = projectRequests.find(
+              (request) =>
+                request.status === 'pending' &&
+                isBefore(addWeeks(new Date(request.createdAt), 2), new Date()),
+            )
+            if (oldPendingRequest) {
+              return 'expired'
+            }
+          }
+          return 'none' // No pending requests found
+        }),
+        catchError((err) => {
+          console.error(err)
+          return of()
+        }),
+      )
+    }
+
+    const subscription = checkRoleRequests$().subscribe({
+      next: (value) => {
+        setLoading(false)
+        setStatus(value)
+      },
+      error: (err) => {
+        console.error(err)
+        setError(err)
+      },
+    })
+
+    return () => {
+      subscription.unsubscribe()
+    }
```

Comment:
Not blocking, this is something we have in some other places of the studio and it works like this.

Only a suggestion, have you tried using the `useObservable` hook here?
It takes care of subscribing and unsubscribing without having to use `useEffect` which makes it simpler.

Also a very nice thing of the observables is that you don't need to push the value to states if you are using the `useObservable` hook given `react-rx` takes care of re rendering the components.

The logic will be the same, the only change you need to do is to return a value with the 3 states:

For example:

```ts
  const checkRoleRequests$ = () => {
      setLoading(true)
      if (!client || !projectId) {
        return of({loading: true, status: undefined, error: null})
      }
    ... rest of the logic, update the returns to match the new objec
  }

 const {loading, status, error} = useObservable(checkRoleRequests$, {loading: true, status: undefined, error:null})
 

```

Reply:
Oh this is so much nicer! Thanks Pedro - updated